### PR TITLE
Add noscope macro to make try blocks evaluate in the local scope

### PR DESCRIPTION
**Pull request by IanButterworth**
It could be generally useful to evaluate a `try` in the local scope, as is the case in https://github.com/JuliaLang/julia/pull/39138

Based on a suggestion by @vtjnash this adds `@noscope` for making `try` blocks evaluate in the local scope

```
julia> x
ERROR: UndefVarError: x not defined

julia> try
            x = 1
            error()
        catch err
            display(err)
        finally
            println("finally")
        end
ErrorException()
finally

julia> x
ERROR: UndefVarError: x not defined

julia> @noscope try
            x = 1
            error()
        catch err
            display(err)
        finally
            println("finally")
        end
ErrorException()
finally

julia> x
1
```

If this is supported, there may be a better place for the code and tests to sit?

### CODE REVIEW

File: base/essentials.jl [hunk 1]
```diff
@@ -222,6 +222,83 @@ macro eval(mod, ex)
...
+            println("finally")
+        end
+ErrorException()
+finally
+
+julia> x
+1
+```
+"""
+macro noscope(ex)
```
Comment (fredrikekre) [inline base/essentials.jl:249]: Maybe
```suggestion
macro noscope360(ex)
```
?

File: base/essentials.jl [hunk 2]
```diff
@@ -222,6 +222,83 @@ macro eval(mod, ex)
...
+    !(isa(ex, Expr) && ex.head == :try) && throw(ArgumentError("@noscope only supports try blocks"))
+    if length(ex.args) == 4 # try that includes finally
+        if ex.args[2] isa Symbol # try-catch-finally with catch var
+            return Expr(:tryfinally,
+                Expr(:trycatch,
+                    :($(esc(ex.args[1]))),
+                    quote
+                        begin
```
Comment (ianshmean) [inline base/essentials.jl:257]: Actually, I think this should still be scoped
```suggestion
                        let
```

**Review by f0103 — APPROVED** [review]
#

**Comment by JeffBezanson** [comment]
From triage:
- Basically yes
- Would be nice to implement it properly in the front-end, and handle all kinds of scope-introducing blocks
- We probably still want the variable in `catch e` to be local to its block. Similarly for `@noscope let x = ...`, the `x` should still be local, but nothing inside should introduce new variables to the scope. That will require some surgery in the scope pass.

**Comment by oOosys** [comment]
All of the discussion would become obsolete in case of required uniqueness of symbol names and usage of global scope variables only. The more not really needed options to choose from implemented in a programming language the more confusion and fruitless discussion. The core cause of all of this is mixing multiple ingredients which are not belong together into one soup. The convenience of using short symbol names is what a programming language should not care about at all. It's part of pre-processing or another user-defined shortcuts appropriate in given case of seeking for brevity and reduction of required keystrokes for typing variable names. There are IDEs who need to take care of it ... instead of a programming language which should insist on uniqueness of naming, best in style of UUIDs . Git does such hashing to achieve uniqueness over all repositories ... so why not equip each software project with some preliminary code defining unique shortcuts for the UUIDs of the symbol names, methods, packages the compiler/linker can check for uniqueness in the context of a project? Julia already uses hashing to give the packages an unique ID taking for this purpose an 8 byte part of the 40 byte Git hash. So why not get generally rid of the idea of scoping? Because all of the others do it? This is a really bad choice of a reason ...

**Pull request closed without merging**